docs/proposal/circuit_breaker.md

```diff
@@ -0,0 +1,374 @@
+            return;
+        }
+
+        bpf_strncpy(data->cluster_name, BPF_DATA_MAX_LEN, (char *)cluster_name);
+        BPF_LOG(DEBUG, KMESH, "record sock bind for cluster %s done\n", cluster_name);
+    }
+    ```
+
++ PASSIVE/ACTIVE ESTABLISHED
+
+    Here, the tcp connection has been established. We can check whether the current socket targets a cluster. If so, we should increase cluster connection counter here.
+
+    We can call the function here:
+
+    ```c
+    static inline void on_cluster_sock_connect(struct bpf_sock_ops *ctx)
+    {
+        if (!ctx) {
+            return;
+        }
+        struct cluster_sock_data *data = get_cluster_sk_data(ctx->sk);
+        if (!data) {
+            return;
+        }
+        __u64 cookie = bpf_get_netns_cookie(ctx);
+        struct cluster_stats_key key = {0};
+        key.netns_cookie = cookie;
+        key.cluster_id = data->cluster_id;
+        BPF_LOG(
+            DEBUG,
+            KMESH,
+            "increase cluster active connections(netns_cookie = %lld, cluster id = %ld)",
+            key.netns_cookie,
+            key.cluster_id);
+        update_cluster_active_connections(&key, 1);
+        BPF_LOG(DEBUG, KMESH, "record sock connection for cluster id = %ld\n", data->cluster_id);
+    }
+    ```
+
++ TCP CLOSE
+
+    Once the tcp connection is closed. We should decrease the counter:
+
+    ```c
+    static inline void on_cluster_sock_close(struct bpf_sock_ops *ctx)
+    {
+        if (!ctx) {
+            return;
+        }
+        struct cluster_sock_data *data = get_cluster_sk_data(ctx->sk);
+        if (!data) {
+            return;
+        }
+        __u64 cookie = bpf_get_netns_cookie(ctx);
+        struct cluster_stats_key key = {0};
+        key.netns_cookie = cookie;
+        key.cluster_id = data->cluster_id;
+        update_cluster_active_connections(&key, -1);
+        BPF_LOG(
+            DEBUG,
+            KMESH,
+            "decrease cluster active connections(netns_cookie = %lld, cluster id = %ld)",
+            key.netns_cookie,
+            key.cluster_id);
+        BPF_LOG(DEBUG, KMESH, "record sock close for cluster id = %ld", data->cluster_id);
+    }
+    ```
+
+We can get the circuit breaker information from cluster data:
+```c
+static inline Cluster__CircuitBreakers *get_cluster_circuit_breakers(const char *cluster_name)
+{
+    const Cluster__Cluster *cluster = NULL;
+    cluster = map_lookup_cluster(cluster_name);
+    if (!cluster) {
+        return NULL;
+    }
+    Cluster__CircuitBreakers *cbs = NULL;
+    cbs = kmesh_get_ptr_val(cluster->circuit_breakers);
+    if (cbs != NULL)
+        BPF_LOG(DEBUG, KMESH, "get cluster's circuit breaker: max connections = %ld\n", cbs->max_connections);
+    return cbs;
+}
+```
+Then, we can get all the thresholds from `Cluster__CircuitBreakers`, and determine whether the circuit breaker should open.
+
+#### Implement the outlier detection function
```

Comment:
Is this the goal of the proposal?

Reply:
yes, I used to store resource limit/thresholds like max_connections in bpf map. But I think direct read them from cluster data would be better, since the space in kernel is limited, no need to store duplicate data.

Reply:
I haven't implemented connection reject now. If we want to reject a connection, we first judge whether some stats exceed the thresholds(we can read them from map_of_cluster).